File: Leetcode/152. Maximum Product Subarray/152_Maximum_Product_Subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int result = nums[0];
        int current_max = 1, current_min = 1;
        for (int n : nums) {
            int temp = n * current_max;
            current_max = Math.max(temp, Math.max(n * current_min, n));
            current_min = Math.min(temp, Math.min(n * current_min, n));
            result = Math.max(current_max, result);
        }
        return result;
    }

    public static void main(String[] args) {
        int[] nums = { 2, 3, -2, 4 };
        Solution s = new Solution();
        int result = s.maxProduct(nums);
        System.out.println(result);
    }
}

File: Leetcode/152. Maximum Product Subarray/152_Maximum_Product_Subarray.py
from typing import List


class Solution:
    def maxProduct(self, nums: List[int]) -> List[int]:
        result = nums[0]
        current_max, current_min = 1, 1

        for n in nums:
            temp = n * current_max
            current_max = max(temp, n * current_min, n)
            current_min = min(temp, n * current_min, n)
            result = max(result, current_max)
        return result


class Main:
    def main(self):
        nums = [2, 3, -2, 4]
        solution = Solution()
        result = solution.maxProduct(nums)
        print(result)


if __name__ == "__main__":
    Main().main()
